### PR TITLE
TranscodeSub

### PR DESCRIPTION
**Pull request by j0sh**
Squashed version of https://github.com/livepeer/go-livepeer-basicnet/pull/32

Code/contents should otherwise be identical; verify with `git diff ja/transcodesub ja/transcodesub-squashed`

### CODE REVIEW

File: basic_network.go [hunk 1]
```diff
@@ -666,7 +689,7 @@ func handleCancelSubReq(nw *BasicVideoNetwork, cr CancelSubMsg, rpeer peer.ID) e
 				if err := nw.sendMessageWithRetry(r.UpstreamPeer, ns, CancelSubID, cr); err != nil {
 					glog.Errorf("Error relaying cancel message to %v: %v ", peer.IDHexEncode(r.UpstreamPeer), err)
 				}
-				return nil
+				return nil // not sure why we exit here; relayer doesn't get cleaned up
```
Comment (ericxtang) [inline basic_network.go]: Oh yeah good point... maybe we shouldn't exit here?

**Review by ericxtang — APPROVED** [review]
🚢 after fixing the old bug in `handleCancelSubReq`.  p.s. the test result inconsistency may come from the `sleep(time.Millisecond*50)` - I find the circleCI environment typically VERY slow, and can cause issues with things like sleep.

**Comment by j0sh** [comment]
Fixed the old bug with cancel here https://github.com/livepeer/go-livepeer-basicnet/pull/38/commits/a9bf5fad8f94ff4a7a9c4edc77d22aec6672cecf (note that the relay mapping key was incorrect as well). Note that I kept the commit standalone since it's unrelated to the transcodesub changes.

As of now, the global `go test` suite doesn't even actually pass locally; it is failing on another older test for no discernible reason. If running each test-file manually, eg `make basic_network_test && make integration_test && make libp2p_playground_test && make network_node_test && make data_test && make transcode_sub_test` then things *usually* pass (but not always). I think the overall issue is less with CircleCI slowness and more to do with inherent issues either in libp2p, our tests, or something in between. This is something we need to address at some point.

**Comment by ericxtang** [comment]
@j0sh for the short term, let's change the circle script to run each file manually.

**Comment by j0sh** [comment]
Updated to use the new libp2p, however circle is failing once again. The same test command works locally.

**Comment by j0sh** [comment]
CircleCI appears to be passing now, but not much changed. 🎉 ❓

**Comment by j0sh** [comment]
Added another fix for a separate basicnet crash that was found during testing. https://github.com/livepeer/go-livepeer-basicnet/pull/38/commits/34c59028b4e80d9e897700d8b6ccaa5e26d37edd

**Comment by ericxtang** [comment]
This is a pretty big PR, and the network architecture has changed since we made this design.  I want to do a little manual testing on Rinkeby before having confidence in it.

**Pull request merged**